### PR TITLE
Add `paginate_req_perform_continue()`

### LINKED ISSUE #298: Store intermediate responses in `req_paginate()`
E.g. after an error or after user interrupt

### PR DESCRIPTION
**Pull request by mgirlich**
Closes #298.

``` r
devtools::load_all("~/GitHub/httr2/")
#> ℹ Loading httr2
page_size <- 7

req_pokemon <- request("https://pokeapi.co/api/v2/pokemon") %>%
  req_url_query(limit = page_size) %>%
  req_verbose(header_resp = FALSE) %>% 
  req_paginate_next_url(
    next_url = function(resp) resp_body_json(resp)[["next"]],
    n_pages = function(resp) {
      total <- resp_body_json(resp)$count
      ceiling(total / page_size)
    }
  )

responses <- paginate_req_perform(req_pokemon, max_pages = 2)
#> -> GET /api/v2/pokemon?limit=7 HTTP/2
#> -> Host: pokeapi.co
#> -> User-Agent: httr2/0.2.3.9000 r-curl/5.0.2 libcurl/8.1.2
#> -> Accept: */*
#> -> Accept-Encoding: deflate, gzip
#> -> 
#> -> GET /api/v2/pokemon?offset=7&limit=7 HTTP/2
#> -> Host: pokeapi.co
#> -> User-Agent: httr2/0.2.3.9000 r-curl/5.0.2 libcurl/8.1.2
#> -> Accept: */*
#> -> Accept-Encoding: deflate, gzip
#> ->

# oh, I actually wanted 3 pages, not 2
responses_all <- paginate_req_perform_continue(max_pages = 3)
#> -> GET /api/v2/pokemon?offset=14&limit=7 HTTP/2
#> -> Host: pokeapi.co
#> -> User-Agent: httr2/0.2.3.9000 r-curl/5.0.2 libcurl/8.1.2
#> -> Accept: */*
#> -> Accept-Encoding: deflate, gzip
#> ->
```

<sup>Created on 2023-09-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/httr2-package.R [hunk 1]
```diff
@@ -15,3 +15,8 @@ the$cache_throttle <- list()
 the$token_cache <- new_environment()
 the$last_response <- NULL
 the$last_request <- NULL
+the$last_pagination_request <- NULL
```
Comment (mgirlich) [inline R/httr2-package.R:18]: Maybe this should go into a list variable?

```r
the$last_pagination <- list(
  request = NULL,
  responses = NULL,
  page = NULL,
  n_pages = NULL,
  max_pages = NULL
)
```

File: R/paginate.R [hunk 2]
```diff
@@ -86,48 +91,99 @@ paginate_req_perform <- function(req,
...
+    resp = resp
+  )
+}
 
+paginate_perform_other_pages <- function(start, n_pages, req, resp) {
   cli::cli_progress_bar(
     "Paginate",
     total = n_pages,
-    format = "{cli::pb_spin} Page {cli::pb_current}/{cli::pb_total} | ETA: {cli::pb_eta}",
-    current = 1L
```
Comment (mgirlich) [inline R/paginate.R:105]: `current` is actually not the current page as I though but whether this is the current progress bar or not.

File: R/paginate.R [hunk 3]
```diff
@@ -86,48 +91,99 @@ paginate_req_perform <- function(req,
...
+    the$last_pagination_responses <- the$last_pagination_responses[seq2(1, page)]
   }
 
-  out
+  the$last_pagination_responses
+}
+
+#' @export
+#' @rdname last_response
+last_pagination_responses <- function() {
```
Comment (mgirlich) [inline R/paginate.R:157]: Should we also export `last_pagination_page()`?

File: R/paginate.R [hunk 4]
```diff
@@ -56,6 +56,11 @@ req_paginate <- function(req,
 
 #' Perform a paginated request
 #'
+#' After preparing a paginated request with [req_paginate()] perform it with
+#' [paginate_req_perform()]. In case the requests were interrupted, you can
```
Comment (hadley) [inline R/paginate.R:60]: Would it be better to copy the `multi_req_perform()` interface here? i.e. `paginate_req_perform()` should always succeed but sometimes return an error object? Or maybe in this case it's sufficient to return a shorter list than expected with a warning.

You could use the same technique as in `pool_run()` to handle user interrupts.

**Comment by hadley** [comment]
Superseded by #353

**Pull request closed without merging**